### PR TITLE
Enable persistent compilation caching

### PR DESCRIPTION
**Pull request by jonb377**
This change enables persistent caching by combining #5800 and #5803. It uses the serialization functionality in PjRtLoadedExecutable to convert the executables to/from strings, which are written to disk by the persistent cache.

The persistent cache is enabled by setting the environment variable `XLA_PERSISTENT_CACHE_PATH` to the desired compilation cache path. An additional environment variable `XLA_PERSISTENT_CACHE_READ_ONLY` can be used to control whether the cache is readonly, which can be useful when the cache is shared across workers in an SPMD setting.

Note that the persistent cache does not perform any eviction, so it is currently up to the user to clean the cache.

### CODE REVIEW

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (jonb377) [inline torch_xla/csrc/runtime/pjrt_computation_client.cc:626]: @JackCaoG I'm not sure what this should be for SPMD mode. Since it's in the runtime layer, this will result in `TPU:0` with SPMD, but should it be `SPMD:0`? It doesn't seem to be causing any issues as is.

**Comment by jonb377** [comment]
Ah bummer, it's not running CI since I'm targeting a different branch... I was hoping to see results for GPU. I guess we'll need to wait until the others land so I can target master.

File: test/test_persistent_cache.py
Comment (JackCaoG) [inline test/test_persistent_cache.py:46]: why not cpu?

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (JackCaoG) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc:626]: Hmm I guess for SPMD we only load it once in one process? I don't remember exactually what I did but I am wondering if we compile the program normally without cache, would device return in `PjRtComputation` has `TPU:0` or `SPMD:0`?

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.cpp:98]: why do we need to check current default device type?

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.cpp:83]: I would prefer to just call be `READ_ONLY` instead of `RO`

File: torch_xla/csrc/xla_graph_executor.h
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.h:169]: I am confuse about the difference between `ComputationCache ` and `MemoryCache `

File: test/run_tests.sh
Comment (JackCaoG) [inline test/run_tests.sh:211]: Thanks @jonb377! I think we want to test 3 different cases
1. single device
2. multi-device non spmd
3. multi-device spmd
and in later time we want to at least manually test in pod.

File: test/test_persistent_cache.py
Comment (jonb377) [reply test/test_persistent_cache.py:46]: I tested it, but CPU isn't supported (deserialization fails). JAX has a similar restriction: https://github.com/google/jax/blob/234be736c4cdd8da4197078278d35a6a1cde3767/tests/compilation_cache_test.py#L69C41-L69C41

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (jonb377) [reply torch_xla/csrc/xla_graph_executor.cpp:98]: I'm a bit outdated on how to check if we're using SPMD mode in the C++ layer - I think this should be the equivalent of what used to be `ShardingUtil::UseVirtualDevice`, but let me know if I'm mistaken.

File: torch_xla/csrc/xla_graph_executor.h
Comment (jonb377) [reply torch_xla/csrc/xla_graph_executor.h:169]: `ComputationCache` is used for the umbrella return type, and `MemoryCache` and `PersistentCache` are subtypes of `ComputationCache`

File: test/run_tests.sh
Comment (jonb377) [reply test/run_tests.sh:211]: Sure! Right now, this covers 1 and 2. I'll add an MP test to the suite.

File: torch_xla/csrc/xla_graph_executor.h
Comment (jonb377) [reply torch_xla/csrc/xla_graph_executor.h:169]: The types were split out in https://github.com/pytorch/xla/pull/5800

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (jonb377) [reply torch_xla/csrc/xla_graph_executor.cpp:98]: Turns out we still have a utility for this! Thanks @JackCaoG 

https://github.com/pytorch/xla/blob/6e661304e270c6a1a864700e04d5d44235c43f3b/torch_xla/csrc/device.cpp#L101-L105

File: test/run_tests.sh
Comment (jonb377) [reply test/run_tests.sh:211]: Just a heads up, the tests are pretty slow since we have to boot up the runtime multiple times (right now it's ~80seconds, with MP it will be 2min).

File: torch_xla/csrc/runtime/pjrt_computation_client.cc
Comment (jonb377) [reply torch_xla/csrc/runtime/pjrt_computation_client.cc:626]: Looks like `SPMD:0`, I'll change this.

File: test/run_tests.sh
Comment (jonb377) [reply test/run_tests.sh:211]: Done - the tests now include:
- Single-device
- Replicated SPMD
- Sharded SPMD
- MP

**Comment by jonb377** [comment]
> @jonb377 do you still want to merge this one?

@JackCaoG Yes, just pending reviews. I'll ping some folks.

File: test/test_persistent_cache.py
Comment (will-cromar) [inline test/test_persistent_cache.py:102]: This looks like a perfect use case for parameterized tests https://github.com/abseil/abseil-py/blob/main/absl/testing/parameterized.py

File: test/test_persistent_cache.py
Comment (will-cromar) [inline test/test_persistent_cache.py:32]: Why does this test case live in a string? If it has to be a separate file, can it be a file checked into this repository? 

If it's possible, I would actually rather you untangle the different test cases into functions and launch them directly, either with xmp.spawn or normal `multiprocessing`, rather than writing a file to run with subprocess. That may be slightly repetitive, but that's okay for unit tests -- it's better to be very explicit about test assertions than build the best abstraction https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html

File: test/test_persistent_cache.py
Comment (will-cromar) [inline test/test_persistent_cache.py:54]: TIL `dict` implements the `<=` operator

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (will-cromar) [inline torch_xla/csrc/xla_graph_executor.cpp:87]: nit: can you fill in these `auto`s? In particular, I can't tell what type `computation` is supposed to be

File: test/test_persistent_cache.py
Comment (jonb377) [reply test/test_persistent_cache.py:102]: Which part should be parametrized?

Comment (jonb377) [reply test/test_persistent_cache.py:102]: Actually - I'm just going to rework the entire test when splitting it out to a separate file.

**Review by yeounoh — APPROVED** [review]
LGTM

File: test/test_persistent_cache.py
Comment (jonb377) [reply test/test_persistent_cache.py:102]: It looks quite nice as a parametrized test!

File: test/test_persistent_cache.py
Comment (jonb377) [reply test/test_persistent_cache.py:32]: I decided to directly use `xmp.spawn` and run SPMD and single-device tests using a ProcessPoolExecutor, so the code lives directly in the test file now. Much cleaner!

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (jonb377) [reply torch_xla/csrc/xla_graph_executor.cpp:87]: I learned while filling in the `auto`s that I also didn't know what type `computation` was lol

File: test/test_persistent_cache.py
Comment (will-cromar) [inline test/test_persistent_cache.py:59]: small suggestion: you can move these test cases into your test class by making them `@staticmethod`s

**Review by will-cromar — APPROVED** [review]
LGTM. Thanks!

File: test/test_persistent_cache.py
Comment (jonb377) [reply test/test_persistent_cache.py:59]: I tried that initially, but for the ProcessPoolExecutor it complained that they weren't pickleable 🥲

Comment (will-cromar) [reply test/test_persistent_cache.py:59]: Hmm, are you sure you didn't use `@classmethod`? We have other tests that use `@staticmethod` to create pickleable test cases

Comment (jonb377) [reply test/test_persistent_cache.py:59]: Just retried the single device test:

```
class PersistentCacheTest(parameterized.TestCase):
  """
  Test suite to verify compilation cache across processes. Tests will run
  multiple Python subprocesses which use the XLA runtime to populate the cache
  and perform assertions on the metrics generated.
  """

  @staticmethod
  def _single_device_test(metrics):
    t = torch.randn(16)
    xt = t.to(xm.xla_device())
    _assert_correctness_and_metrics(t, xt, metrics)
```

It hit this:

```
Traceback (most recent call last):
  File "/home/ptxla/.local/lib/python3.8/site-packages/absl/testing/parameterized.py", line 321, in bound_param_test
    return test_method(self, *testcase_params)
  File "test_persistent_cache.py", line 23, in run
    f(*args, **kwargs)
  File "test_persistent_cache.py", line 103, in test_persistent_cache
    launch_method(test_fn, ({
  File "test_persistent_cache.py", line 31, in _test_spawn
    pool.submit(fn, *args).result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.8/multiprocessing/queues.py", line 239, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/local/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle 'staticmethod' object
```

The _mp_test also hit this when using staticmethod... I wonder what's going wrong lol

**Comment by jonb377** [comment]
MP GPU looks good, but single-device seems to be broken by the same issue addressed in #6023 (cc @vanbasten23). I've disabled single-device and SPMD GPU tests.

**Pull request merged**
